cpu/stm32/periph/usbdev_fs.c

```diff
@@ -0,0 +1,661 @@
+            return USB_EP_ISOCHRONOUS;
+        case USB_EP_TYPE_BULK:
+            return USB_EP_BULK;
+        case USB_EP_TYPE_INTERRUPT:
+            return USB_EP_INTERRUPT;
+        default:
+            assert(false);
+            return 0;
+    }
+}
+
+void usbdev_init_lowlevel(void)
+{
+    for (size_t i = 0; i < USBDEV_NUMOF; i++) {
+        _setup(&_usbdevs[i], &stm32_usbdev_fs_config[i]);
+    }
+}
+
+/*TODO: rework this function (maybe move it outside this file)
+        for other ST families support */
+static void _enable_usb_clk(void)
+{
+    /* Validate USB Supply */
+    PWR->CR2 |= PWR_CR2_USV;
+    RCC->APB1SMENR1 |= RCC_APB1SMENR1_USBSMEN;
+    /* Reset CSR to default settings */
+    RCC->CSR |= RCC_CSR_RFRSTS;
+    RCC->CSR &= ~RCC_CSR_RFRSTS;
+    /* Enable CRS with auto trim enabled */
+    CRS->CR |=  CRS_CR_AUTOTRIMEN |  CRS_CR_CEN;
+}
+
+static void _enable_gpio(const stm32_usbdev_fs_config_t *conf)
+{
+    gpio_init(conf->dp, GPIO_IN);
+    gpio_init(conf->dm, GPIO_IN);
+    /* Configure AF for the pins */
+    gpio_init_af(conf->dp, conf->af);
+    gpio_init_af(conf->dm, conf->af);
+}
+
+static void _set_ep_in_status(uint16_t *val, uint16_t mask)
+{
+    /* status endpoints bits can only be toggled, writing 0
+       as no effect. Thus, check which bits should be set
+       then XOR it with their current value */
+    if (mask & USB_EPTX_DTOG1) {
+        *val ^= USB_EPTX_DTOG1;
+    }
+    if (mask & USB_EPTX_DTOG2) {
+        *val ^= USB_EPTX_DTOG2;
+    }
+}
+
+static void _set_ep_out_status(uint16_t *val, uint16_t mask)
+{
+       /* status endpoints bits can only be toggled, writing 0
+       as no effect. Thus, check which bits should be set
+       then XOR it with their current value */
+    if (mask & USB_EPRX_DTOG1) {
+        *val ^= USB_EPRX_DTOG1;
+    }
+    if (mask & USB_EPRX_DTOG2) {
+        *val ^= USB_EPRX_DTOG2;
+    }
+}
+
+static inline void _usb_attach(stm32_usbdev_fs_t *usbdev)
+{
+    /* Some ST USB IP doesn't have this feature */
+#ifdef USB_BCDR_DPPU
+    const stm32_usbdev_fs_config_t *conf = usbdev->config;
+    bool syncing = false;
+    /* Enable DP pullup to signal connection */
+    _global_regs(conf)->BCDR |= USB_BCDR_DPPU;
+    do {
+        syncing = CRS->ISR & 0x00000008;
+    } while (!syncing);
+#else
+    (void)usbdev;
+#endif /* USB_BCDR_DPPU */
+}
+
+static inline void _usb_detach(stm32_usbdev_fs_t *usbdev)
+{
+    /* Some ST USB IP doesn't have this feature */
+#ifdef USB_BCDR_DPPU
+    const stm32_usbdev_fs_config_t *conf = usbdev->config;
+    DEBUG_PUTS("usbdev_fs: Detaching from host");
+
+    /* Disable DP pullup to signal disconnection */
+    CLRBIT(_global_regs(conf)->BCDR, USB_BCDR_DPPU);
+#else
+    (void)usbdev;
+#endif
+}
+
+static void _set_address(stm32_usbdev_fs_t *usbdev, uint8_t address)
+{
+    const stm32_usbdev_fs_config_t *conf = usbdev->config;
+    _global_regs(conf)->DADDR = USB_DADDR_EF | (address & USB_DADDR_ADD);
+}
+
+void isr_usb_lp(void) {
```

Comment:
on L0, the USB IRQ subroutine is called `isr_usb` (computed based on the IRQ enum) so to avoid CPU family related issues, what about adding a define in the `periph_conf.h`file and use it here ?
Something like:

- in `periph_conf.h`:

```c
#define USBDEV_ISR              isr_usb
```

- here:

```c
void USBDEV_ISR(void) {
```

Reply:
Done.